### ISSUE TITLE
Need to analyse Columbia database for performance enhancements in AIP creation

### ISSUE THREAD

**Issue by kellyannewithane** [artefactual-labs/archivematica-acceptance-tests#62]
After analysing how micro-services work in parallel when creating an AIP in a test environment, now need to determine how those parallels play out using Columbia data

![am_task_parallelism_1024](https://user-images.githubusercontent.com/25352922/34738589-4f75e1e2-f52e-11e7-840a-2f6b81a928ee.png)

**Comment by jrwdunham** [comment]
The graph below shows the creation of an AIP from the sampledata Images/ directory. On a machine with 4 CPUs, a total of 472.2 seconds of micro-service task processing time was performed in 139.26 seconds, a compression rate of 0.3. This indicates that 4 micro-service tasks were executing in parallel at nearly all times.

![images-aip-concurrency](https://user-images.githubusercontent.com/2504271/34806791-f8a8aa90-f639-11e7-8f27-d305e14973c0.png)

**Comment by jrwdunham** [comment]
The graphs in this issue were created using the data from the Tasks table after creating AIPs using the automated procedure made possible by https://github.com/artefactual-labs/archivematica-acceptance-tests/pull/47. Building on this performance testing analysis, future work could involve gathering timing information like this from inputs (transfers) more representative of the Columbia data and generating the same types of graphs in order to discover low-hanging fruit where performance enhancement efforts should be focused.

**Comment by jrwdunham** [comment]
Performance Analysis Report
================================================================================

**This is a draft.**

Columbia wants Archivematica to process transfers/ingests more quickly and to
build artifacts (e.g., METS files) that are not so large that they are unusable
or slow down processing.

This is a report on the findings of recent work to analyze the performance of
Archivematica. This analysis has taken the following form:

1. Timing of transfers with and without full output stream inter-process
   communication.
2. SQL queries on Columbia db dump.
3. Graphical view of parallel task execution.
4. Profiling of AM processes (pyflame and flame graphs)


Timing of transfers with and without full output streams
================================================================================

Result: discovered that a 6% reduction in total processing time can be achieved
simply by ceasing to serialize and then pass full standard output and standard
error strings from MCP client (gearman worker) processes over the wire to the
MCP server (gearman scheduler).

The above result was arrived at by writing and running the
performance-stdout-no-write.feature feature file. This feature file describes a
replicable experiment, which in short form is as follows:

1. Create an AIP with an Archivematica deploy configured to PASS output streams
   between processes.
2. Perform SQL queries to retrieve and store task durations, i.e., the time
   between when the MCP server (scheduler) submits a task and when it receives
   notice of that task's completion from the MCP client.
3. Create an AIP with an Archivematica deploy configured to NOT pass output
   streams between processes.
4. Perform SQL queries to retrieve and store task durations.
5. Compute overall task processing time in the "yes output stream" and "no
   output stream" conditions and compare the two, expecting to find that the
   latter condition results in a decreased total task processing time.

See https://github.com/artefactual-labs/archivematica-acceptance-tests/blob/master/features/core/performance-stdout-no-write.feature


SQL queries on Columbia db dump.
================================================================================

Justin executed a number of SQL queries on an Archivematica database dump
provided to us by Columbia. The target AIP originated from a transfer of 10,000
image files at 457G total size (up to 160M in size, but most under 10M). The
URL to the CUL Archivematica Performance Queries google doc is
https://docs.google.com/document/d/1Fv8SthGKndeIg_TzVNNXXJUg89Rm0PYC80ebf50gIuI/edit.

The primary take-away from these measurements is that avoidable micro-services
should be avoided.

- Most costly micro-service groups (can contain multiple costly micro-services!)

  - Normalize 58 hours!! (but it includes norm for access, preservation and
    thumbnails, so really 19 hours for each normalize)

    - some normalization can be avoided, e.g., creation of thumbnails for XML
      files.

  - METS XML generation/update 47 hours

    - Update METS XML took 27 hours!

      - why all the METS generation and updating? Can we avoid some of this?
      - how can we reduce the size of the METS files?
      - how can we improve performance here?

    - Generate AIP METS 12.5 hours
    - Generate transfer METS 7.5 hours

  - Clean up names took 32 hours!

    - python logic and writing to disk intermixed. Some smarter programming
      could mean real efficiencies here probably.

  - Extract packages took a day (12 hours)

    - maxmize true multi-core/multi-node parallelism here.

  - Verify checksums 2 hours

    - maybe there is some needless redundancy here with "Prepare AIP"s BagIt
      work.

  - Identify file format 1 hour

    - the transfer/ingest redundancy could be avoided by inspecting the FPRules
      for changes.

- One apparent false reading is Create SIP from Transfer, which appears to have
  taken 113 hours, but it did not.

  - Specific micro-services was "Check transfer directory for objects".

    - It actually occurred 3 times, with 1 second duration witha 5-day span
      between the first and third occurrences, so it was a measurement error.

- Another false positive is "Verify transfer compliance" (micro-service "Verify
  mets_structmap.xml compliance") which appears to have taken 0 cpu time but 41
  hours wall time.


- We should avoid having MCP clients spawn a new Python process, which imports and
  sets up Django, just for the new process to determine that it has no work to
  do, and then do that 30K times, once for each file, which we do.

  - For some long-running tasks, this may be allowable.
  - E.g., a .tiff file will probably have no preservation normalization
    rule---discover this in the long-running process with the one-time start-up
    cost.

We need to discover the specific trade-offs between short-lived python
multi-process task distribution, long-lived micro-services, true thread-based
parallelism, and sequential (non-concurrent) iteration are.

Python + Django import and setup = ~1.25 seconds (for normalize.py).

Current state:

- 1 MCP Client process

  - 4+ Python threads

    - each fork (block on, and terminate) a process, a client script (~1 second cost)

      - may fork a process, an FPR command

Possible state:

- Multi-threaded MCP Client process(es) as above, but with suitable client
  scripts registered.

- 4+ MCP Client processes (have Django imported and set-up)

  - no threads
  - long-lived process (orchestrated docker container)
  - import client scripts (Python re-importing has negligible performance cost)
  - no 1s python/django spin-up cost
  - client script may fork a process, an FPR command, e.g., Ffmpeg

- Clojure/GoLang MCP client (gearman worker)

  - true multi-threaded (multi-core) parallelism for CPU-bound work
  - make every core on your cluster work
  - Use Onyx or Storm

    - Onyx could do non-microservice pipeline work

Opinion alert. We should stop using watched directories and we should start doing event
sourcing. Archivematica is not just a preservation pipeline, it is a
knowledge system. The ultimate source of truth of Archivematica's state cannot
be a disjoint collection of non-indexed METS XML files on disk. Archivematica
should process data, persist it for a time, and persist events permanently, or
rather beliefs about events, which is to say, knowledge.


Graphical view of parallel task execution.
================================================================================

See https://github.com/artefactual-labs/archivematica-acceptance-tests/issues/62#issuecomment-356811271

Task compression through concurrency (parallelization) is 93% of optimal. That
is, say, 100 hours of total task time will take 30 hours on a 4-cpu system when
optimal would be 25 hours (70/75 == 0.93).

  - adding more cpus/nodes for a higher parallelism factor will obviously help.
  - there are 7 obvious gaps in the processing where no tasks are being performed

    - how much of the total AIP processing time does this represent? From
      visual inspection of the graph it looks like 1/8th or 12.5%.
    - my guess is that this is moving lots of files between watched directories.

  - some overhead is unavoidable, but increasing the concurrency task time
    compression rate (e.g., via long-lived processes) will have real effects.


Profiling of AM processes (pyflame and flame graphs)
================================================================================

PyFlame is a sampling profiler: it inspects the call stack of a Python process
and shows the lines of code (calls) that took the most total time.

- MCPServer and MCPClient processes would appear to have no obvious and
  significant performance optimizations.
- Individual client scripts (e.g., "normalize.py") would appear to be spending
  roughly 1.25 seconds in needlessly forking new Python processes, importing
  Django (and other heavy packages), and setting up Django.

  - This could be avoided by:

    - importing client scripts into the long-running Python (MCP Client)
      process instead of forking new Python processes.
    - giving client scripts the state (db data, settings, unit variables) they
      need, and receiving commands to mutate the db from the client scripts.
      Outsource state managemet from short-lived client scripts to long-lived
      MCP Client processes.

**Comment by jrwdunham** [comment]
@cul-archivematica @jhsimpson I am closing this because AM and the Columbia sample database have been analyzed for performance improvements as shown in the comments above as well as the internal Google docs entitled "Columbia Phase 2.1 Report" and "CUL Archivematica Performance Queries". Next steps in terms of performance-enhancing sub-projects have been decided upon based on these analyses and new "Columbia"-tagged GH issues have been created and are being worked on.
